### PR TITLE
Allow to override db name using AWS tag.

### LINKED ISSUE #12446: Teleport, with auto-discovery only detects one DB if there are multiple DBs with same instanceID (in different regions/accounts)
Description:

With `rds autodiscovery` where DB InstanceID is same across different regions, only one db is showing up in Teleport.

Expected behavior:

All DB instances (with auto-discovery) should show up in teleport, even if they have same InstanceID. These DB names can be changed to custom names or prefix/suffix some labels to identify specific Instance.

Current behavior:
Only one DB shows up in Teleport, which ever is discovered first.

Bug details:
- Teleport version: 9.1.2
- Recreation steps

#### Setup:

* DB instance `postgres01` created in both `ap-southeast-2` and `ap-southeast-1` and DBinstance `testpgdb` is created in `ap-southeast-1`.
* db-agent node is running in `ap-southeast-2`.

```
[root@ip-172-31-4-12 teleport]# for region in ap-southeast-2 ap-southeast-1
> do
> echo $region
> aws rds describe-db-instances --region=$region --query 'DBInstances[].DBInstanceIdentifier[]'
> done
ap-southeast-2
[
    "postgres01"
]
ap-southeast-1
[
    "postgres01",
    "testpgdb"
]
```

#### Scenario 1:  `regions` setting as below in `/etc/teleport.yaml` of db-agent. i.e. `ap-southeast1` is first in the list, db instance `postgres01` from `ap-southeast-1` shows in teleport.

``` 
aws:
  # RDS/Aurora databases auto-discovery.
  # For more information about RDS/Aurora auto-discovery: https://goteleport.com/docs/database-access/guides/rds/
  - types: ["rds"]
    # AWS regions to register databases from.
    regions:
    - ap-southeast-1
    - ap-southeast-2
    # AWS resource tags to match when registering databases.
    tags:
      "*": "*"
```

``` 
% tsh db ls
Name       Description                    Labels                                                                                                   Connect
---------- ------------------------------ -------------------------------------------------------------------------------------------------------- -------
postgres01 RDS instance in ap-southeast-1 account-id=xxxxxxxxxxxx,endpoint-type=instance,engine-version=13.4,engine=postgres,region=ap-southeast-1
testpgdb   RDS instance in ap-southeast-1 account-id=xxxxxxxxxxxx,endpoint-type=instance,engine-version=13.4,engine=postgres,region=ap-southeast-1
```

#### Scenario 2:  `regions` setting as below in `/etc/teleport.yaml` of db-agent. i.e. `ap-southeast2` is first in the list, db instance `postgres01` from `ap-southeast-2` shows in teleport.

```  
aws:
  # RDS/Aurora databases auto-discovery.
  # For more information about RDS/Aurora auto-discovery: https://goteleport.com/docs/database-access/guides/rds/
  - types: ["rds"]
    # AWS regions to register databases from.
    regions:
    - ap-southeast-2
    - ap-southeast-1
    # AWS resource tags to match when registering databases.
    tags:
      "*": "*"
```

``` 
% tsh db ls
Name       Description                    Labels                                                                                                   Connect
---------- ------------------------------ -------------------------------------------------------------------------------------------------------- -------
postgres01 RDS instance in ap-southeast-2 account-id=xxxxxxxxxxxx,endpoint-type=instance,engine-version=13.4,engine=postgres,region=ap-southeast-2
testpgdb   RDS instance in ap-southeast-1 account-id=xxxxxxxxxxxx,endpoint-type=instance,engine-version=13.4,engine=postgres,region=ap-southeast-1
```

#### Scenario 3:  if I start two separate processes one each for separate region (on same db-agent node), `tctl` shows db is there but `tsh` doesnt show the DB.

``` 
% tctl db ls
Host                                              Name       Protocol URI                                               Labels                                                                                                   Version
------------------------------------------------- ---------- -------- ------------------------------------------------- -------------------------------------------------------------------------------------------------------- -------
ip-172-31-4-12.ap-southeast-2.compute.internal    postgres01 postgres postgres01.cgvd4dfsnp9h.ap-southeast-1.rds.ama... account-id=xxxxxxxxxxxx,endpoint-type=instance,engine-version=13.4,engine=postgres,region=ap-southeast-1 9.1.3
ip-172-31-4-12.ap-southeast-2.compute.internal    testpgdb   postgres testpgdb.cgvd4dfsnp9h.ap-southeast-1.rds.amazo... account-id=xxxxxxxxxxxx,endpoint-type=instance,engine-version=13.4,engine=postgres,region=ap-southeast-1 9.1.3
ip-172-31-4-12.ap-southeast-2.compute.internal... postgres01 postgres postgres01.c1qpy0jaasz7.ap-southeast-2.rds.ama... account-id=xxxxxxxxxxxx,endpoint-type=instance,engine-version=13.4,engine=postgres,region=ap-southeast-2 9.1.3

 % tsh db ls
Name       Description                    Labels                                                                                                   Connect
---------- ------------------------------ -------------------------------------------------------------------------------------------------------- -------
postgres01 RDS instance in ap-southeast-1 account-id=xxxxxxxxxxxx,endpoint-type=instance,engine-version=13.4,engine=postgres,region=ap-southeast-1
testpgdb   RDS instance in ap-southeast-1 account-id=xxxxxxxxxxxx,endpoint-type=instance,engine-version=13.4,engine=postgres,region=ap-southeast-1

 %
```

- Debug logs
Attached DB agents logs. for scenario1 and scenario2.
[teleport-db-agent-logs-1.txt](https://github.com/gravitational/teleport/files/8629134/teleport-db-agent-logs-1.txt)
[teleport-db-agent-logs-2.txt](https://github.com/gravitational/teleport/files/8629135/teleport-db-agent-logs-2.txt)

### PR DESCRIPTION
**Pull request by Tener**
* If `teleport.dev/database_name` tag is present, it will override automatic DB name.
* If a single AWS DB maps to several Teleport DB entries, the override will become part of the name.

Among other things, this change makes it possible to pick a different name in case there are multiple dbs with the same name present.

Fixes https://github.com/gravitational/teleport/issues/12446

Note: As of https://github.com/gravitational/teleport/pull/14922 `teleport.dev/database-name` is changed to  `teleport.dev/database_name`

### CODE REVIEW

**Review by smallinsky — COMMENTED** [review]
Regarding https://github.com/gravitational/teleport/pull/14740#issuecomment-1191548518
I agree,  we should make the tag formatting consistent: 
The `teleport-db-name` will introduce third formatting style that we are using so we either can make this consistent with EC2 tag `TeleportHostname` and use `TeleprotDBName` tag  or apply here the `teleport.dev/databse-name` convention.

**Comment by Tener** [comment]
> Regarding [#14740 (comment)](https://github.com/gravitational/teleport/pull/14740#issuecomment-1191548518) I agree, we should make the tag formatting consistent: The `teleport-db-name` will introduce third formatting style that we are using so we either can make this consistent with EC2 tag `TeleportHostname` and use `TeleprotDBName` tag or apply here the `teleport.dev/databse-name` convention.

I think I like the `teleport.dev/database-name` version better.

File: lib/services/database.go [hunk 1]
```diff
@@ -118,6 +118,16 @@ func UnmarshalDatabase(data []byte, opts ...MarshalOption) (types.Database, erro
 	return nil, trace.BadParameter("unsupported database resource version %q", h.Version)
 }
 
+// newDatabase is a wrapper around types.NewDatabaseV3 that additionally applies tag-based name override.
+// For some database types we don't want to override the entire name; if this is desired, caller will pass
+// suffix to be appended to the database name.
+func newDatabase(suffix string, meta types.Metadata, spec types.DatabaseSpecV3) (types.Database, error) {
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/services/database.go:124]: I tried several options before settling on the `newDatabase` wrapper function. 

The problem is that we need to inspect a different field (labels) and modify the name accordingly. In several cases, there is no variable for `labels` present so we'd need to introduce it, which makes the code a bit harder to read and opens up the possibility for mistakes.

Still, the fact that you find the `newDatabase` call confusing is good feedback; I'll try to improve it for clarity.

[the remark this answers is not in the document]
Comment (Tener) [reply lib/services/database.go]: Done: https://github.com/gravitational/teleport/pull/14740/commits/caf393e6db488db10262394d467b57ac4b8cd22b

**Review by greedy52 — APPROVED** [review]

File: lib/services/database.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (greedy52) [reply lib/services/database.go]: Sigh. It's like chicken and egg. I think both the original way and the new way are OK.

Just my 2c. If we are ok introducing a new struct to make it slightly clearer, probably can do something like
```
    return createDatabase(createDatabaseParam{
        DefaultName:  aws.StringValue(instance.DBClusterIdentifier),
        NameSuffixes: []string{string(RDSEndpointTypeReader)},
        Metadata: types.Metadata{
            Description: fmt.Sprintf("Aurora cluster in %v (%v endpoint)", metadata.Region, string(RDSEndpointTypeReader)),
            Labels:      labelsFromRDSCluster(cluster, metadata, RDSEndpointTypeReader),
        },
        Spec: types.DatabaseSpecV3{
            Protocol: engineToProtocol(aws.StringValue(cluster.Engine)),
            URI:      fmt.Sprintf("%v:%v", aws.StringValue(cluster.ReaderEndpoint), aws.Int64Value(cluster.Port)),
            AWS:      *metadata,
        },
    })

**Review by smallinsky — APPROVED** [review]

File: lib/services/database.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tener) [reply lib/services/database.go]: Eh, I think this style of code may simply not be Go's strongest point. 

F# has a nice feature where you can easily copy an entire struct but change a few chosen fields:

```fsharp
let myRecord3 = { myRecord2 with Y = 100; Z = 2 }
```

https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/copy-and-update-record-expressions

I think I'll keep the current code for now...

**Comment by r0mant** [comment]
@Tener Did we add any docs for this? We should if we haven't yet.

**Comment by Tener** [comment]
> @Tener Did we add any docs for this? We should if we haven't yet.

@r0mant not yet, but I'm working on it.

**Pull request merged**